development-practices/defining-the-boundaries-of-a-package.md

```diff
@@ -0,0 +1,107 @@
+
+**CCP** amplifies this by grouping together classes components cannot be closed against certain types of changes into the same packages. Thus, when a change in requirements comes along; that change has a good chance of being restricted to a minimal number of packages.
+
+### Principles on Package Coupling
+
+#### The Acyclic Dependencies Principle (ADP)
+
+> **The dependency structure between packages must be a directed acyclic graph (DAG). That is there must be no cycles in the dependency structure**
+
+Cyclical dependencies are not allowed in packages for a reason, it makes deploying packages impossible, imagine the following scenario where one has three packages, A, B and C, with the following dependency order, where B is dependent on A, C is dependent on B and A, and A itself which depends on Package C
+
+![Package C with a dependency cycle](../.gitbook/assets/package-dependency.png)
+
+These packages are impossible to deploy on a new Org. One can't deploy Package A without deploying Package C and Package C itself can't be deployed without deploying Package A resulting in a standoff. Hence, cycles should be broken either by inverting the dependencies or by introducing a shared package which shares the common components for both Package A and Package C.
+
+#### The Stable Dependencies Principle (SDP)
+
+> **The dependencies between packages in a design should be in the direction of the stability of the packages. A package should only depend upon packages that are more stable that it is.**
+
+Designs cannot be completely static. Some volatility is necessary if the design is to be maintained. We accomplish this by conforming to the _Common Closure Principle_ (**CCP**). By using this principle, we create packages that are sensitive to certain kinds of changes. These packages are designed to be volatile. We expect them to change.
+
+Any package that we expect to be volatile should not be depended upon by a package that is difficult to change! Otherwise the volatile package will also be difficult to change.
+
+By conforming to the **SDP**, we ensure that modules that are designed to be instable (i.e. easy to change) are not depended upon by modules that are more stable (i.e. harder to change) than they are.
+
+#### The Stable Abstractions Principle (SAP) \[Code-Related]
+
+> **Packages that are maximally stable should be maximally abstract. Instable packages should be concrete. The abstraction of a package should be in proportion to its stability.**
+
+This principle sets up a relationship between stability and abstractness. It says that a stable package should also be abstract so that its stability does not prevent it from being extended. On the other hand, it says that an instable package should be concrete since its instability allows the concrete code within it to be easily changed.
```

Comment:
```suggestion
This principle sets up a relationship between stability and abstractness. It says that a stable package should also be abstract so that its stability does not prevent it from being extended. On the other hand, it says that an unstable package should be concrete since its instability allows the concrete code within it to be easily changed.
```